### PR TITLE
Adjoint method for Euler solver

### PR DESCRIPTION
**Pull request by gautierronan**

### CODE REVIEW

**Review by abocquet — APPROVED** [review]

File: dynamiqs/mesolve/euler.py [hunk 1]
```diff
@@ -1,10 +1,17 @@
...
 from .me_solver import MESolver
 
 
-class MEEuler(MESolver, FixedSolver):
+class MEEuler(MESolver, AdjointFixedSolver):
     def forward(self, t: float, rho: Tensor) -> Tensor:
         # rho: (b_H, b_rho, n, n) -> (b_H, b_rho, n, n)
         return rho + self.dt * self.lindbladian(t, rho)
+
+    def backward_augmented(self, t: float, rho: Tensor, phi: Tensor) -> Tensor:
```
Comment (pierreguilmin) [inline dynamiqs/mesolve/euler.py:12]:
```suggestion
    def backward_augmented(self, t: float, rho: Tensor, phi: Tensor) -> tuple[Tensor, Tensor]:
```
Should also be fixed in `rouchon.py` and `fixed_solver.py`.

Comment (pierreguilmin) [inline dynamiqs/mesolve/euler.py:12]:
```suggestion
    def backward_augmented(self, t: float, rho: Tensor, phi: Tensor) -> tuple[Tensor, Tensor]:
```
Should also be fixed in `rouchon.py` and `fixed_solver.py`.

File: dynamiqs/mesolve/me_solver.py [hunk 2]
```diff
@@ -26,3 +26,14 @@ def lindbladian(self, t: float, rho: Tensor) -> Tensor:
         H = self.H(t)
         out = -1j * self.H_nh(H) @ rho + 0.5 * kraus_map(rho, self.jump_ops)
         return out + out.mH  # (b_H, b_rho, n, n)
+
+    def adjoint_lindbladian(self, t: float, rho: Tensor) -> Tensor:
+        """Compute the action of the adjoint Lindbladian on the density matrix.
```
Comment (pierreguilmin) [inline dynamiqs/mesolve/me_solver.py:31]:
```suggestion
        """Compute the action of the adjoint Lindbladian on an operator.
```

File: dynamiqs/mesolve/me_solver.py [hunk 3]
```diff
@@ -26,3 +26,14 @@ def lindbladian(self, t: float, rho: Tensor) -> Tensor:
         H = self.H(t)
         out = -1j * self.H_nh(H) @ rho + 0.5 * kraus_map(rho, self.jump_ops)
         return out + out.mH  # (b_H, b_rho, n, n)
+
+    def adjoint_lindbladian(self, t: float, rho: Tensor) -> Tensor:
```
Comment (pierreguilmin) [inline dynamiqs/mesolve/me_solver.py:30]:
```suggestion
    def adjoint_lindbladian(self, t: float, x: Tensor) -> Tensor:
```

**Review by pierreguilmin — CHANGES_REQUESTED** [review]
Nice! Minor comments. It would be nice if you could do a PR for the Schrödinger Euler solver after too.

**Review by pierreguilmin — CHANGES_REQUESTED** [review]
Nice! Minor comments. It would be nice if you could do a PR for the Schrödinger Euler solver after too.

**Comment by pierreguilmin** [comment]
Related to #192.

File: dynamiqs/mesolve/me_solver.py [hunk 4]
```diff
@@ -26,3 +26,14 @@ def lindbladian(self, t: float, rho: Tensor) -> Tensor:
...
+        """Compute the action of the adjoint Lindbladian on the density matrix.
+
+        Notes:
+            Hermiticity of the output is enforced to avoid numerical instability
+            with Runge-Kutta solvers.
+        """
+        H = self.H(t)
+        out = 1j * self.H_nh(H).mH @ rho + 0.5 * kraus_map(rho, self.jump_ops.mH)
```
Comment (pierreguilmin) [inline dynamiqs/mesolve/me_solver.py:38]:
```suggestion
        out = 1j * self.H_nh(H) @ rho + 0.5 * kraus_map(rho, self.jump_ops.mH)
```

**Comment by gautierronan** [comment]
> It would be nice if you could do a PR for the Schrödinger Euler solver after too.

I realized it's not so simple. Plus I don't expect anyone will every be using that. I don't know, something to think about.

**Review by pierreguilmin — DISMISSED** [review]
Very nice 🥳

**Pull request merged**